File: elements-of-programming-interviews/2023/python/is_string_palindromic.py
from test_framework import generic_test


def is_palindromic(s: str) -> bool:
    if not s:
        return False
    l = 0
    r = len(s)-1
    while l <= r:
        if s[l] != s[r]:
            return False
        l += 1
        r -= 1
    return True


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('is_string_palindromic.py',
                                       'is_string_palindromic.tsv',
                                       is_palindromic))

File: elements-of-programming-interviews/2023/cpp/is_string_palindromic.cc
#include <string>

#include "test_framework/generic_test.h"
using std::string;

bool IsPalindromic(const string& s) {
  if (s.empty()) {
    return false;
  }
  int l = 0, r = s.size() - 1;
  while (l <= r) {
    if (s[l++] != s[r--]) {
      return false;
    }
  }

  return true;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"s"};
  return GenericTestMain(args, "is_string_palindromic.cc",
                         "is_string_palindromic.tsv", &IsPalindromic,
                         DefaultComparator{}, param_names);
}
